packages/mobile/src/escrow/saga.ts

```diff
@@ -117,6 +217,125 @@ function* registerStandbyTransaction(context: TransactionContext, value: string,
   )
 }
 
+async function formEscrowWithdrawTxWithNoCode(
+  contractKit: ContractKit,
+  escrowWrapper: EscrowWrapper,
+  paymentId: string,
+  pepper: string,
+  msgHash: string,
+  addressIndex: number
+) {
+  const { privateKey } = generateDeterministicInviteCode(pepper, addressIndex)
+  const signature: string = (await contractKit.web3.eth.accounts.sign(msgHash, privateKey))
+    .signature
+  Logger.debug(
+    TAG + '@withdrawFromEscrowViaKomenci',
+    `Signed message hash signature is ${signature}`
+  )
+
+  const { r, s, v } = splitSignature(contractKit, signature)
+  const withdrawTx = escrowWrapper.withdraw(paymentId, v, r, s)
+  return withdrawTx
+}
+
+function* withdrawFromEscrowUsingPepper(komenciActive: boolean = false) {
+  try {
+    ValoraAnalytics.track(OnboardingEvents.escrow_redeem_start)
+    Logger.debug(TAG + '@withdrawFromEscrowViaKomenci', 'Withdrawing escrowed payment')
+    const phoneHashDetails: PhoneNumberHashDetails | undefined = yield call(
+      getUserSelfPhoneHashDetails
+    )
+
+    if (!phoneHashDetails) {
+      throw Error('Couldnt find own phone hash or pepper. Should never happen.')
+    }
+
+    const { phoneHash, pepper } = phoneHashDetails
+    const [contractKit, walletAddress, mtwAddress, feelessVerificationState]: [
+      ContractKit,
+      string,
+      string,
+      FeelessVerificationState
+    ] = yield all([
+      call(getContractKit),
+      call(getConnectedUnlockedAccount),
+      select(mtwAddressSelector),
+      select(feelessVerificationStateSelector),
+    ])
+
+    const komenciKit: KomenciKit = new KomenciKit(contractKit, walletAddress, {
+      url: KOMENCI_URL,
+      token: feelessVerificationState.komenci.sessionToken,
+    })
+
+    const escrowWrapper: EscrowWrapper = yield call([
+      contractKit.contracts,
+      contractKit.contracts.getEscrow,
+    ])
+
+    const escrowPaymentIds: string[] = yield call(
+      [escrowWrapper, escrowWrapper.getReceivedPaymentIds],
+      phoneHash
+    )
+
+    const msgHash = contractKit.web3.utils.soliditySha3({ type: 'address', value: walletAddress })
+    const context = newTransactionContext(TAG, 'Withdraw from escrow')
+    // TODO: Batch the tranasctions and submit them together via `executeTransactions`
+    // method on an instance of the MTW then submitting like usual
+    const withdrawTxSuccess: boolean[] = []
+    for (let i = 0; i < escrowPaymentIds.length; i += 1) {
+      const withdrawTx: CeloTransactionObject<boolean> = yield formEscrowWithdrawTxWithNoCode(
+        contractKit,
+        escrowWrapper,
+        escrowPaymentIds[i],
+        pepper,
+        msgHash,
+        i
+      )
+
+      try {
+        if (!komenciActive) {
+          yield call(sendTransaction, withdrawTx.txo, walletAddress, context)
```

Comment:
Should we have a balance check before getting to this point? In order to redeem the escrow without Komenci, they will need funds in their account already. The effect of this is that this transaction would fail at gas estimation.

Reply:
I believe the assumption is that if they have enough funds needed to complete verification, then they will have enough remaining to pay for gas for this tx.

I agree that this is not a resilient approach. Given the breadth of changes, my goal was to achieve UX parity so, while it is a valid concern, I'm inclined to tackle this in the larger rework needed for the escrow redemption process within Valora.